worker/volume_sweeper.go

```diff
@@ -0,0 +1,98 @@
+package worker
+
+import (
+	"context"
+	"os"
+	"sync"
+	"time"
+
+	"code.cloudfoundry.org/lager"
+	"code.cloudfoundry.org/lager/lagerctx"
+	"github.com/concourse/baggageclaim"
+)
+
+// volumeSweeper is an ifrit.Runner that periodically reports and
+// garbage-collects a worker's volumes
+type volumeSweeper struct {
+	logger             lager.Logger
+	interval           time.Duration
+	tsaClient          TSAClient
+	baggageclaimClient baggageclaim.Client
+	maxInFlight        int
+}
+
+func NewVolumeSweeper(
+	logger lager.Logger,
+	sweepInterval time.Duration,
+	tsaClient TSAClient,
+	bcClient baggageclaim.Client,
+	maxInFlight int,
+) (*volumeSweeper) {
+	return &volumeSweeper{
+		logger:             logger,
+		interval:           sweepInterval,
+		tsaClient:          tsaClient,
+		baggageclaimClient: bcClient,
+		maxInFlight:        maxInFlight,
+	}
+}
+
+func (sweeper *volumeSweeper) Run(signals <-chan os.Signal, ready chan<- struct{}) error {
+	timer := time.NewTicker(sweeper.interval)
+
+	close(ready)
+
+	for {
+		select {
+		case <-timer.C:
+			sweeper.sweep(sweeper.logger.Session("tick"))
+
+		case sig := <-signals:
+			sweeper.logger.Info("sweep-cancelled-by-signal", lager.Data{"signal": sig})
+			return nil
+		}
+	}
+}
+
+func (sweeper *volumeSweeper) sweep(logger lager.Logger) {
+	ctx := lagerctx.NewContext(context.Background(), logger)
+
+	volumes, err := sweeper.baggageclaimClient.ListVolumes(logger.Session("list-volumes"), baggageclaim.VolumeProperties{})
+	if err != nil {
+		logger.Error("failed-to-list-volumes", err)
+	} else {
+		handles := []string{}
+		for _, volume := range volumes {
+			handles = append(handles, volume.Handle())
+		}
+
+		err := sweeper.tsaClient.ReportVolumes(ctx, handles)
+		if err != nil {
+			logger.Error("failed-to-report-volumes", err)
+		}
+	}
+
+	volumeHandles, err := sweeper.tsaClient.VolumesToDestroy(ctx)
+	if err != nil {
+		logger.Error("failed-to-sweep-volumes", err)
```

Comment:
Should we change this action to something else? Seems a bit vague for what could actually be failing here 🤔

Reply:
It would mostly be a connection error of some sort, but that error would be surfaced as well. The log line would have something like `failed-to-sweep-volumes, err: <actual_err_msg>`. I think the `failed-to-sweep-volumes` is just to give context on where the following error occurred.

Reply:
Updated the log messages to add handles. Is this better? Wdyt?

Reply:
hmmmm I see, I was looking at it from the perspective that the error would come from the fact that *listing the volume handles to be destroyed* failed, instead of the whole action of sweeping them 🤔 

e.g, consider this panel:

<img width="1075" alt="screen shot 2019-03-07 at 12 04 28 pm" src="https://user-images.githubusercontent.com/3574444/53974730-3d29c180-40d1-11e9-8ee5-35a4e05191c5.png">

for something like that, I imagine that being able to see the more specific action that is failing (getting the list of volumes to destroy) can help troubleshooting more than a more generic one.

Wdyt?

Reply:
Changed the error message in both container- and volume-sweeper to reflect that it's the listing volumes/containers that fails. Does this make it clearer?

Reply:
oooh it does! thank you!!